### PR TITLE
mcobbett unit testing couchdb

### PR DESCRIPTION
**Pull request by techcobweb**
- initial unit test added, validator of db separated
- dont checkin vscode settings
- bump release.yaml version of couchdb code
- tidy copyright and unused imports
- mocks moved to subfolder. Testcase can create a ras store
- feature flag can turn on/off behviour
- unit tests capture HTTP interactions. Versions always rise with release.

The overall effect of these code changes is to enable more unit testing of what http traffic flows back and forth between the extension and couchdb.

These changes should not effect the actual functioning of the behaviour at all.

This PR contributes to the issue [Spike: CouchDB : Single binary stored in single couchdb document#1653](https://github.com/galasa-dev/projectmanagement/issues/1653)

### CODE REVIEW

**Comment by galasa-team** [comment]
Build successful

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/CouchdbValidatorImplTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    //     MockCloseableHttpClient mockHttpClient = new MockCloseableHttpClient(response);
+
+    //     CouchdbValidator validatorUnderTest = new CouchdbValidatorImpl();
+
+    //     // When..
+    //     Throwable thrown = catchThrowable(()-> validatorUnderTest.checkCouchdbDatabaseIsValid( rasURI , mockHttpClient ));
+
+    //     // Then..
+    //     assertThat(thrown).isInstanceOf(CouchdbRasException.class);
+    // }
```
Comment (eamansour) [inline galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/CouchdbValidatorImplTest.java:48]: Do we want this test or more tests for the CouchdbValidatorImpl class?

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/CouchdbTestFixtures.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            response.setEntity(entity);
+
+            return response;
+        }
+    }
+
+
+    // public static class CreateArtifactDocument extends BaseHttpInteraction {
+    //     // http://my.uri/galasa_artifacts
+    // }
```
Comment (eamansour) [inline galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/CouchdbTestFixtures.java:180]: Not needed?

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockCloseableHttpClient.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            this.currentInteraction = interactionWalker.next();
+        } else {
+             this.currentInteraction = null ;
+        }
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public HttpParams getParams() {
+        // TODO Auto-generated method stub
```
Comment (eamansour) [inline galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockCloseableHttpClient.java:45]: Very minor but I'd remove these TODOs - same applies in other mocks in this PR

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockHttpEntity.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public InputStream getContent() throws IOException, UnsupportedOperationException {
+        return new ByteArrayInputStream(this.payloadMessageBytes);
+    }
+
+    @Override
+    public long getContentLength() {
+        return this.payloadMessageBytes.length;
+    }
```
Comment (eamansour) [inline galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockHttpEntity.java:78]: Might be good to move the implemented methods above the ones that haven't been implemented

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/CouchdbValidatorImplTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import org.junit.Rule;
+import org.junit.rules.TestName;
+
+public class CouchdbValidatorImplTest {
+    
+    @Rule
+    public TestName testName = new TestName();
+
+
+    // @Test
```
Comment (techcobweb) [inline galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/CouchdbValidatorImplTest.java:17]: This test needs reviving now that the other tests work.

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockCloseableHttpResponse.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+
+    @Override
+    public HttpEntity getEntity() {
+        return this.entity;
+    }
+
+    @Override
+    public Locale getLocale() {
+        // TODO Auto-generated method stub
```
Comment (techcobweb) [inline galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockCloseableHttpResponse.java:34]: Remove the TODOs

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockFramework.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+import dev.galasa.framework.spi.IRun;
+import dev.galasa.framework.spi.SharedEnvironmentRunType;
+import dev.galasa.framework.spi.creds.CredentialsException;
+import dev.galasa.framework.spi.creds.ICredentialsService;
+
+public class MockFramework implements IFramework {
+
+    @Override
+    public void setFrameworkProperties(Properties overrideProperties) {
+        // TODO Auto-generated method stub
```
Comment (techcobweb) [inline galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockFramework.java:35]: Remove the TODO markers

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockHeaderElement.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public NameValuePair getParameter(int arg0) {
+        // TODO Auto-generated method stub
```
Comment (techcobweb) [inline galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockHeaderElement.java:28]: Remove the TODO markers

**Review by eamansour — APPROVED** [review]
A few minor suggestions but happy to approve

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockFramework.java (hunk 7 above)
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockFramework.java]: fixed

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockHeaderElement.java (hunk 8 above)
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockHeaderElement.java]: fixed

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockConfigurationPropertyStoreService.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            throws ConfigurationPropertyStoreException {
+        String key = prefix + "." + suffix ;
+        String value = props.get(key);
+        return value ;
+    }
+
+    @Override
+    public @NotNull Map<String, String> getPrefixedProperties(@NotNull String prefix)
+            throws ConfigurationPropertyStoreException {
+        // TODO Auto-generated method stub
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockConfigurationPropertyStoreService.java]: fixed

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockCloseableHttpResponse.java (hunk 6 above)
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockCloseableHttpResponse.java]: fixed

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockCloseableHttpClient.java (hunk 3 above)
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockCloseableHttpClient.java]: all fixed

**Comment by galasa-team** [comment]
Build successful

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockHttpEntity.java (hunk 4 above)
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/MockHttpEntity.java:35]: Agreed. I created a base class which implemented no methods at all, inherited from that, and it's very clear now which methods are implemented with no generated garbage mixed in. Fixed.

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/CouchdbTestFixtures.java (hunk 2 above)
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/mocks/CouchdbTestFixtures.java:180]: Correct. Not neeed. Removed.

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

File: galasa-extensions-parent/dev.galasa.cps.etcd/build.gradle [hunk 10]
```diff
@@ -25,4 +25,5 @@ dependencies {
        force = true
     }
     implementation ('com.google.guava:failureaccess:1.0.1')
+    implementation ('dev.galasa:dev.galasa.framework: 0.31.0')
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.cps.etcd/build.gradle:28]: How does that work. Perhaps you can explain tomorrow...
The thing is, the etcd code relies on a version of the framework, as it draws interfaces from there no ? So there is a dependency, so when framework bumps up, then etcd should reflect that also no ?

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/CouchdbValidatorImplTest.java (hunk 5 above)
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/CouchdbValidatorImplTest.java:17]: same comment as below. Resolving that one and dismissing this one.

File: galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/CouchdbValidatorImplTest.java (hunk 1 above)
Comment (techcobweb) [reply galasa-extensions-parent/dev.galasa.ras.couchdb/src/test/java/dev/galasa/ras/couchdb/internal/CouchdbValidatorImplTest.java]: Got this one working again. Testing this stuff properly, could take ages, and we will now add tests which at least test the code we are about to change.

**Comment by galasa-team** [comment]
Build successful

**Pull request merged**